udp/udp.go

```diff
@@ -0,0 +1,187 @@
+package udpconn
+
+import (
+	"fmt"
+	"github.com/nknorg/encrypted-stream"
+	"github.com/nknorg/tuna/pb"
+	"io"
+	"net"
+	"sync"
+)
+
+const (
+	PrefixLen            = 4
+	DefaultUDPBufferSize = 38324
+	MaxUDPBufferSize     = 65527
+)
+
+type Conn interface {
+	WriteMsgUDP(b, oob []byte, addr *net.UDPAddr) (n, oobn int, err error)
+	ReadFromUDP(b []byte) (n int, addr *net.UDPAddr, err error)
+
+	Close() error
+
+	LocalAddr() net.Addr
+	RemoteAddr() net.Addr
+
+	SetWriteBuffer(bytes int) error
+	SetReadBuffer(bytes int) error
+}
+
+type EncryptUDPConn struct {
+	conn Conn
+
+	encoders map[string]*stream.Encoder
+	decoders map[string]*stream.Decoder
+
+	lock     sync.RWMutex
+	isClosed bool
+
+	readLock  sync.Mutex
+	writeLock sync.Mutex
+
+	readBuffer    []byte
+	writeBuffer   []byte
+	decryptBuffer []byte
+	encryptBuffer []byte
+}
+
+func NewEncryptUDPConn(conn Conn) *EncryptUDPConn {
+	ec := &EncryptUDPConn{
+		conn:          conn,
+		encoders:      make(map[string]*stream.Encoder),
+		decoders:      make(map[string]*stream.Decoder),
+		readBuffer:    make([]byte, MaxUDPBufferSize),
+		writeBuffer:   make([]byte, MaxUDPBufferSize),
+		encryptBuffer: make([]byte, MaxUDPBufferSize),
+		decryptBuffer: make([]byte, MaxUDPBufferSize),
+	}
+	return ec
+}
+
+func (ec *EncryptUDPConn) AddCodec(addr *net.UDPAddr, encryptKey *[32]byte, encryptionAlgo pb.EncryptionAlgo, initiator bool) error {
+	var cipher stream.Cipher
+	var err error
+	switch encryptionAlgo {
+	case pb.EncryptionAlgo_ENCRYPTION_NONE:
+		return nil
+	case pb.EncryptionAlgo_ENCRYPTION_XSALSA20_POLY1305:
+		cipher = stream.NewXSalsa20Poly1305Cipher(encryptKey)
+	case pb.EncryptionAlgo_ENCRYPTION_AES_GCM:
+		cipher, err = stream.NewAESGCMCipher(encryptKey[:])
+		if err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("unsupported encryption algo %v", encryptionAlgo)
+	}
+	encoder, err := stream.NewEncoder(cipher, initiator, true)
+	if err != nil {
+		return err
+	}
+	decoder, err := stream.NewDecoder(cipher, initiator, true, true)
+	if err != nil {
```

Comment:
Same here, and we can enable nonce verification because there is no compatibility issue